src/ECL/ECL.hs

```diff
@@ -43,15 +56,38 @@ data Guard = CatchAll | Opaque
 -- a collection of values, and an optional guard.
 type Alternative lit = (Binders lit, Maybe Guard)
 
--- | A list of uncovered cases
-newtype Uncovered lit = Uncovered { getUncovered :: [Binders lit] }
-  deriving (Show, Eq) 
+-- | Check wraps both uncovered and redundant cases
+data Check lit = Check
+  { getUncovered :: [Binders lit]
+  , getRedundant :: (Maybe Bool, [Binders lit])
+  } deriving (Show, Eq) 
+
+makeCheck :: [Binders lit] -> (Maybe Bool, [Binders lit]) -> Check lit
+makeCheck bs red = Check
+  { getUncovered = bs
+  , getRedundant = red }
+
+unwrapCheck :: Check lit -> ([Binders lit], (Maybe Bool, [Binders lit]))
+unwrapCheck ch = (getUncovered ch, getRedundant ch)
+
+applyUncovered :: ([Binders lit] -> [Binders lit]) -> Check lit -> Check lit
+applyUncovered f c = makeCheck (f $ getUncovered c) (getRedundant c)
+
+applyRedundant :: ((Maybe Bool, [Binders lit]) -> (Maybe Bool, [Binders lit])) -> Check lit -> Check lit
+applyRedundant f c = makeCheck (getUncovered c) (f $ getRedundant c)
+
+applyCheck ::
```

Comment:
I'd probably just use composition if I needed both. Don't force yourself to use it if you don't need to.